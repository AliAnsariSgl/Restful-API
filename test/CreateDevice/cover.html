
<!DOCTYPE html>
<html>
	<head>
		<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
		<style>
			body {
				background: black;
				color: rgb(80, 80, 80);
			}
			body, pre, #legend span {
				font-family: Menlo, monospace;
				font-weight: bold;
			}
			#topbar {
				background: black;
				position: fixed;
				top: 0; left: 0; right: 0;
				height: 42px;
				border-bottom: 1px solid rgb(80, 80, 80);
			}
			#content {
				margin-top: 50px;
			}
			#nav, #legend {
				float: left;
				margin-left: 10px;
			}
			#legend {
				margin-top: 12px;
			}
			#nav {
				margin-top: 10px;
			}
			#legend span {
				margin: 0 5px;
			}
			.cov0 { color: rgb(192, 0, 0) }
.cov1 { color: rgb(128, 128, 128) }
.cov2 { color: rgb(116, 140, 131) }
.cov3 { color: rgb(104, 152, 134) }
.cov4 { color: rgb(92, 164, 137) }
.cov5 { color: rgb(80, 176, 140) }
.cov6 { color: rgb(68, 188, 143) }
.cov7 { color: rgb(56, 200, 146) }
.cov8 { color: rgb(44, 212, 149) }
.cov9 { color: rgb(32, 224, 152) }
.cov10 { color: rgb(20, 236, 155) }

		</style>
	</head>
	<body>
		<div id="topbar">
			<div id="nav">
				<select id="files">
				
				<option value="file0">DevicesServ\test\CreateDevice\main.go (89.8%)</option>
				
				</select>
			</div>
			<div id="legend">
				<span>not tracked</span>
			
				<span class="cov0">not covered</span>
				<span class="cov8">covered</span>
			
			</div>
		</div>
		<div id="content">
		
		<pre class="file" id="file0" style="display: none">package main

import (
        model "DevicesServ/DataModel"
        responses "DevicesServ/Responses"
        "encoding/json"
        "errors"
        "net/http"
        "os"

        "github.com/aws/aws-lambda-go/events"
        "github.com/aws/aws-lambda-go/lambda"
        "github.com/aws/aws-sdk-go/aws"
        "github.com/aws/aws-sdk-go/aws/session"
        "github.com/aws/aws-sdk-go/service/dynamodb"
        "github.com/aws/aws-sdk-go/service/dynamodb/dynamodbattribute"
)

var sess *session.Session

// Init function Initialize session
func init() <span class="cov8" title="1">{
        // Initialize a session
        region := os.Getenv("AWS_REGION")
        sess = session.Must(session.NewSession(&amp;aws.Config{Region: &amp;region}))

}</span>

// Check Missing Payloads
func CheckPayloads(PostRequestJSON map[string]interface{}) error <span class="cov8" title="1">{
        Flag := false
        ErrorText := " Incomplete Request:needed payloads:"

        if PostRequestJSON["id"] == nil </span><span class="cov8" title="1">{
                Flag = true
                ErrorText += " 'id' "
        }</span>
        <span class="cov8" title="1">if PostRequestJSON["deviceModel"] == nil </span><span class="cov8" title="1">{
                Flag = true
                ErrorText += " 'deviceModel' "
        }</span>
        <span class="cov8" title="1">if PostRequestJSON["name"] == nil </span><span class="cov8" title="1">{
                Flag = true
                ErrorText += " 'name' "
        }</span>
        <span class="cov8" title="1">if PostRequestJSON["note"] == nil </span><span class="cov8" title="1">{
                Flag = true
                ErrorText += " 'note' "
        }</span>
        <span class="cov8" title="1">if PostRequestJSON["serial"] == nil </span><span class="cov8" title="1">{
                Flag = true
                ErrorText += " 'serial' "
        }</span>

        //  return error
        <span class="cov8" title="1">if Flag == true </span><span class="cov8" title="1">{
                return errors.New(ErrorText)
        }</span>

        <span class="cov8" title="1">return nil</span>
}

// insert request json payload into new device
func makeNewDevice(PayloadJSON map[string]interface{}) model.Device <span class="cov8" title="1">{
        return model.Device{
                ID:          PayloadJSON["id"].(string),
                DeviceModel: PayloadJSON["deviceModel"].(string),
                Name:        PayloadJSON["name"].(string),
                Note:        PayloadJSON["note"].(string),
                Serial:      PayloadJSON["serial"].(string),
        }
}</span>

//Insert Devices to database
func InsertDeviceToDatabase(device model.Device) error <span class="cov8" title="1">{

        // Initialize a session in AWS_REGION Using SDK
        region := os.Getenv("AWS_REGION")
        sess := session.Must(session.NewSession(&amp;aws.Config{Region: &amp;region}))

        // Create DynamoDB client
        svc := dynamodb.New(sess)
        db := mockDynamoDB{svc}
        // Convert Device object to dynamodb AttributeValues
        av, err := dynamodbattribute.MarshalMap(device)
        if err != nil </span><span class="cov0" title="0">{
                return err
        }</span>

        // Insert av into DevicesTable
        <span class="cov8" title="1">DevicesTable := aws.String(os.Getenv("DEVICES_TABLE"))
        _, err = db.PutItem(&amp;dynamodb.PutItemInput{
                Item:      av,
                TableName: DevicesTable,
        })

        if err != nil </span><span class="cov0" title="0">{
                return err
        }</span>

        <span class="cov8" title="1">return nil</span>
}

// Handler function
func Handler(r events.APIGatewayProxyRequest) (events.APIGatewayProxyResponse, error) <span class="cov8" title="1">{
        // Unmarshal request payload
        var PostRequestJSON map[string]interface{}
        if err := json.Unmarshal([]byte(r.Body), &amp;PostRequestJSON); err != nil </span><span class="cov8" title="1">{
                return responses.IncompleteRequest(errors.New("")), nil
        }</span>

        // finding missing payloads
        <span class="cov8" title="1">err := CheckPayloads(PostRequestJSON)
        if err != nil </span><span class="cov8" title="1">{
                return responses.IncompleteRequest(err), nil
        }</span>

        // call method to insert request json payload into new device
        <span class="cov8" title="1">device := makeNewDevice(PostRequestJSON)

        // Insert device object to dynamodb
        if err := InsertDeviceToDatabase(device); err != nil </span><span class="cov0" title="0">{
                return responses.InternalServerError(), nil
        }</span>

        // make response
        <span class="cov8" title="1">deviceJSON, err := json.Marshal(device)
        if err != nil </span><span class="cov0" title="0">{
                return responses.InternalServerError(), nil
        }</span>
        //create new device
        <span class="cov8" title="1">return events.APIGatewayProxyResponse{
                StatusCode: http.StatusCreated,
                Headers:    map[string]string{"Content-Type": "application/json"},
                Body:       string(deviceJSON),
        }, nil</span>
}

func main() <span class="cov0" title="0">{
        lambda.Start(Handler)
}</span>
</pre>
		
		</div>
	</body>
	<script>
	(function() {
		var files = document.getElementById('files');
		var visible;
		files.addEventListener('change', onChange, false);
		function select(part) {
			if (visible)
				visible.style.display = 'none';
			visible = document.getElementById(part);
			if (!visible)
				return;
			files.value = part;
			visible.style.display = 'block';
			location.hash = part;
		}
		function onChange() {
			select(files.value);
			window.scrollTo(0, 0);
		}
		if (location.hash != "") {
			select(location.hash.substr(1));
		}
		if (!visible) {
			select("file0");
		}
	})();
	</script>
</html>
