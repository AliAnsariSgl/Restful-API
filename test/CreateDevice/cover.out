mode: set
DevicesServ\test\CreateDevice\main.go:22.13,27.2 2 1
DevicesServ\test\CreateDevice\main.go:30.66,34.34 3 1
DevicesServ\test\CreateDevice\main.go:38.2,38.43 1 1
DevicesServ\test\CreateDevice\main.go:42.2,42.36 1 1
DevicesServ\test\CreateDevice\main.go:46.2,46.36 1 1
DevicesServ\test\CreateDevice\main.go:50.2,50.38 1 1
DevicesServ\test\CreateDevice\main.go:56.2,56.18 1 1
DevicesServ\test\CreateDevice\main.go:60.2,60.12 1 1
DevicesServ\test\CreateDevice\main.go:34.34,37.3 2 1
DevicesServ\test\CreateDevice\main.go:38.43,41.3 2 1
DevicesServ\test\CreateDevice\main.go:42.36,45.3 2 1
DevicesServ\test\CreateDevice\main.go:46.36,49.3 2 1
DevicesServ\test\CreateDevice\main.go:50.38,53.3 2 1
DevicesServ\test\CreateDevice\main.go:56.18,58.3 1 1
DevicesServ\test\CreateDevice\main.go:64.69,72.2 1 1
DevicesServ\test\CreateDevice\main.go:75.56,86.16 6 1
DevicesServ\test\CreateDevice\main.go:91.2,97.16 3 1
DevicesServ\test\CreateDevice\main.go:101.2,101.12 1 1
DevicesServ\test\CreateDevice\main.go:86.16,88.3 1 0
DevicesServ\test\CreateDevice\main.go:97.16,99.3 1 0
DevicesServ\test\CreateDevice\main.go:105.87,108.73 2 1
DevicesServ\test\CreateDevice\main.go:113.2,114.16 2 1
DevicesServ\test\CreateDevice\main.go:119.2,122.55 2 1
DevicesServ\test\CreateDevice\main.go:127.2,128.16 2 1
DevicesServ\test\CreateDevice\main.go:132.2,136.8 1 1
DevicesServ\test\CreateDevice\main.go:108.73,110.3 1 1
DevicesServ\test\CreateDevice\main.go:114.16,116.3 1 1
DevicesServ\test\CreateDevice\main.go:122.55,124.3 1 0
DevicesServ\test\CreateDevice\main.go:128.16,130.3 1 0
DevicesServ\test\CreateDevice\main.go:139.13,141.2 1 0
